docs/design/audit-v2.md

```diff
@@ -0,0 +1,78 @@
+By choosing nodes out of the reservoirs, we would receive a random sample for auditing.
+
+• We have a reservoir of k items and a stream of n items, where n is an unknown number.
+• Fill the reservoir from [0...k-1] with the first k items in the stream
+• For every item in the stream from index i=k..n-1, pick a random number j=rand(0..i), and if j<k, replace reservoir[j] with stream[i]
+
+## Rationale
+
+An audit observer on the metainfo loop, like every observer on the loop, will always hit every segment on the satellite exactly once, allowing us to get the most accurate possible sample for each node.
+
+While we initially considered integrating the audit system's random node selection process with the existing garbage collection observer, we decided not to do this because the difference in required interval for each observer would mean either too many bloom filters being created unnecessarily or audits occurring too slowly.
+
+An initial idea for implementation was to sort the nodes table for nodes with least amount of audits, then select one node randomly within that low amount of audits.
+However, we decided it may not be necessary to keep track of how many audits per storage node if we're able to randomly select across nodes.
+
+We were also considering implementing a reverse way of looking up segments or pieces by node IDs.
+e.g. a table where each row is a node ID and an encrypted metainfo path. 
+Every time a segment is committed or deleted, that table (and every node) gets updated.
+The advantage is that it can super simplify the garbage collection process, but complexify upload and download.
+We ended up deciding to avoid adding new tables if possible.
+
+## Implementation
+
+1. Add reservoir sampling struct for node auditing.
+2. Create an audit observer that connects to metainfo loop but otherwise uses existing audit code for selection.
+3. Have the audit observer update the reservoir sampling structs.
+3. Make selection biased in favor of unvetted nodes, and select a random segment to audit based on the populated reservoir sampling struct for that node.
+4. Audit the segment selected from part 3, as we are doing already.
+
+## Open issues
+
+1. Location of reservoir sampling. (resolved)
+    - Where do we have this happen? In a garbage collection observer or a new observer?
+
+From Moby: "The main issue with integrating it into the gc observer is it means we will always be forcing the gc interval and the node audit reservoir sampling interval to be exactly the same. I don't think the performance gain of combining the two is necessarily worth the limitations created. Plus, this is the entire reason we created the metainfo loop/observer architecture."
+
+    - We want uploads to be performant with minimal db transactions, but we know that audits need to happen very frequently.
+    - We'll create a new audit observer.
+
+2. Should we run both audit selection processes within the same loop or in separate loops?
```

Comment:
Joining into a loop doesn't create a new iteration unless the timings are different.

We could add a join together that ensures that multiple observers always. But let's keep the logic separate to avoid conflating different approaches.

Reply:
does this mean I need to add further clarification to the doc? if so, what do I need to add?

Reply:
Let's keep the observers separate, it makes the logic easier to read.